File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.py
def solution(info, edges):
    n = len(info)
    visited = [False for _ in range(n)]
    ans = []
    
    def dfs(sheep: int, wolf: int, v):
        if sheep > wolf:
            ans.append(sheep)
            # ans = max(ans, sheep)
        else:
            return
        
        for parent, child in edges:
            if v[parent] and not v[child]:
                v[child] = True
                
                if not info[child]:
                    dfs(sheep + 1, wolf, v)
                else:
                    dfs(sheep, wolf + 1, v)
                    
                v[child] = False
    
    
    visited[0] = True
    dfs(1, 0, visited)
    
    return max(ans)

    

File: 프로그래머스/lv3/92343. 양과 늑대/양과 늑대.java
class Solution {
    boolean[] visited;
    int[] i;
    int[][] e;
    int ans = -1, n = 0;
    public int solution(int[] info, int[][] edges) {
        n = info.length;
        visited = new boolean[n];
        e = new int[edges.length][];
		for (int i = 0; i < n; i++) visited[i] = false;

		i = info.clone();
        for (int i = 0; i < edges.length; i++) e[i] = edges[i].clone();

		visited[0] = true;
		dfs(1, 0);

        return ans;
    }

    void dfs(int sheep, int wolf) {
        if (sheep > wolf) ans = Math.max(ans, sheep);
        else return;

        for (int[] next: e){
            if (visited[next[0]] && !visited[next[1]]) {
                visited[next[1]] = true;

                if (i[next[1]] == 0)    dfs(sheep + 1, wolf);
                else                    dfs(sheep, wolf + 1);

                visited[next[1]] = false;
            }
        }
    }
}